pkg/rpc/context_test.go

```diff
@@ -2655,47 +2514,147 @@ func TestHeartbeatDialback(t *testing.T) {
-	// shut down the reverse connection. Normally the connect attempt times out
-	// immediately and returns an error, but occasionally it needs to wait for the
-	// RPCHeartbeatTimeout (100 ms). Wait until pings have stopped in both
-	// directions for at least 1 second before checking health.
+	// n1 -> n2 can not be healthy for much longer since the listener for n2 is
+	// closed, and n2 should terminate its connection to n1 within ~100ms
+	// (heartbeat interval in this test) since n1 can't dial-back. We'll give a
+	// generous 10s for that to happen, but bail early once we see connections in
+	// both directions marked as unhealthy.
+	tr := timeutil.NewTimer()
+	defer tr.Stop()
+	tr.Reset(10 * time.Second)
 	for {
+		strict := false
+		// After a short moment, should no longer see pings n2->n1 here because
+		// n1's dialback will fail (so they're rejected), even though n2's circuit
+		// breaker will regularly try to reach out.
 		select {
-		case ping := <-pingChan1:
-			log.Infof(ctx, "Received %+v", ping)
-		case ping := <-pingChan2:
-			log.Infof(ctx, "Received %+v", ping)
-		case <-time.After(1 * time.Second):
-			err1 := ctx1.ConnHealth(remoteAddr2, 2, DefaultClass)
-			require.True(t, errors.Is(err1, ErrNoConnection), "%+v", err1)
-			err2 := ctx2.ConnHealth(remoteAddr1, 1, SystemClass)
-			require.True(t, errors.Is(err2, ErrNoConnection), "%+v", err2)
+		case ping := <-nonRejectedPings1:
+			t.Logf("Received ping n%d->n%d", ping.OriginNodeID, ping.TargetNodeID)
+		case ping := <-nonRejectedPings2:
+			t.Logf("Received ping n%d->n%d", ping.OriginNodeID, ping.TargetNodeID)
+		case <-tr.C:
+			tr.Read = true
+			strict = true
+		case <-time.After(time.Millisecond):
+		}
+
+		errDef, errSys := ctx1.ConnHealth(remoteAddr2, 2, DefaultClass), ctx2.ConnHealth(remoteAddr1, 1, SystemClass)
+		if errors.Is(errDef, ErrNotHeartbeated) {
+			errDef = nil
+		}
+		if errors.Is(errSys, ErrNotHeartbeated) {
+			errSys = nil
+		}
+		if errors.HasType(errDef, (*netutil.InitialHeartbeatFailedError)(nil)) {
+			errDef = nil
+		}
+		if errors.HasType(errSys, (*netutil.InitialHeartbeatFailedError)(nil)) {
+			errSys = nil
+		}
+		if errDef == nil && errSys == nil {
 			return
 		}
+
+		if strict {
+			require.NoError(t, errDef)
+			require.NoError(t, errSys)
+		}
 	}
 }
 
 // TODO(baptist): Add a test using TestCluster to verify this works in a full
 // integration test.
+
+func checkMetrics(m *Metrics, healthy, unhealthy int64, checkDurations bool) error {
+	// NB: peers can be neither healthy nor unhealthy. This happens when
+	// a peer is first created and when its deleteAfter field is set.
+	if exp, n := healthy, m.ConnectionHealthy.Value(); exp != n {
+		return errors.Errorf("ConnectionHealthy = %d", n)
+	}
+	if exp, n := unhealthy, m.ConnectionUnhealthy.Value(); exp != n {
+		return errors.Errorf("ConnectionUnhealthy = %d", n)
+	}
+
+	if !checkDurations {
+		return nil
+	}
+
+	if hf := m.ConnectionHealthyFor.Value(); (healthy > 0) != (hf != 0) {
```

Comment:
`(healthy > 0) != (hf != 0)` is not obvious at all. Not sure we can simplify the condition, but at least a comment saying that `checkDurations` mean HealthyFor, UnhealthyFor and RoundTrip must be non zero if corresponding counters are non zero.

Reply:
I added this comment:

	// HealthyFor is nonzero if and only if there is at least one healthy connection, since it's
	// the sum of the connections' HealthyFor, and on a connection this is nonzero if and only
	// if the connection is healthy.

Does that help or do you think further clarification is necessary? The condition is exact, so all we can do is rewrite it. `hf > 0` is probably better than `hf != 0`, I made that change.